### DISCUSSION TITLE
Questions about `noise_sv2` crate

### DISCUSSION THREAD

**Discussion by plebhash** [stratum-mining/stratum#916] [General]
this discussion should guide the documentation of `noise_sv2`, as described on https://github.com/stratum-mining/stratum/issues/845

**Comment by plebhash** [comment]
the [`aes-gcm`](https://crates.io/crates/aes-gcm) crate provides an implementation of `AES-GCM` cipher, which is an alternative to `ChaCha20Poly1305` cipher (provided by [`chacha20poly1305`](https://crates.io/crates/chacha20poly1305) crate)

both crates are used as dependencies for `noise_sv2`

inside `impl AeadCipher for ChaCha20Poly1305` we have:

https://github.com/stratum-mining/stratum/blob/24b8b1d8666d6055c4a964d2aac8ef2445734a10/protocols/v2/noise-sv2/src/aed_cipher.rs#L27-L34

I'm confused: how can `ChaCha20Poly1305::encrypt_in_place` return `aes_gcm::Error`, if they come from different crates?

**Reply by plebhash** [reply]
ah I see: both `aes_gcm::Error` and `chacha20poly1305::Error` are actually re-exports of `aead::Error`:

https://github.com/RustCrypto/AEADs/blob/f4e39243d2d3ca9ceea4c6e1782359ff9400bbb9/chacha20poly1305/src/lib.rs#L146
https://github.com/RustCrypto/AEADs/blob/f4e39243d2d3ca9ceea4c6e1782359ff9400bbb9/aes-gcm/src/lib.rs#L111 

note for the documentation efforts: it is probably be a good idea to replace `aes_gcm::Error` with `aead::Error` to avoid this kind of confusion to other readers

**Comment by bit-aloo** [comment]
@Fi3, In `noise_sv2`, we're implementing `AEAD` ciphers for two algorithms: 1. `ChaCha20Poly1305` and 2. `AES256-GCM`. However, we're only using `ChaCha20Poly1305` for Noise. Is there a specific reason for including `AES256-GCM`?

**Reply by jbesraa** [reply]
They both have the same goal so It does not make sense to keep both. I would keep `ChaCha20Poly1305` as it is the standard in the bitcoin ecosystem.

Edit: @Shourya742 this might be useful https://stratumprotocol.org/specification/04-Protocol-Security/#43-choice-of-cryptographic-primitives

**Reply by Fi3** [reply]
the sepc say that we should have both if I remeber correctly. We could have removed that part btw I should double check

**Reply by Fi3** [reply]
yes we need to support both:
https://github.com/stratum-mining/sv2-spec/blob/main/04-Protocol-Security.md#441-cipherstate-object

**Reply by Fi3** [reply]
if we want to remove we should change the spec. Not against it, but do we really need to remove one? I agree that if we want to keep only one it have to be chachapoly

**Reply by plebhash** [reply]
- Was there some historical reason to support two different flavours of `AEAD` cipher?
- Is there a good argument for keeping `AES256-GCM`?

@TheBlueMatt do you know the answer to these questions? Or maybe should we reach out to the Braiins guys?

Unless there's a compelling reason for keeping it, I would lean towards dropping `AES256-GCM` support (both in spec and SRI).

Otherwise we will waste engineering effort on maintaining something that nobody will ever use.

**Reply by Fi3** [reply]
maintaining it is not a meaningful effort btw I think that if there is no reason to have it we should remove it, maybe @jakubtrnka have a better understanding of why we have it?

**Reply by jakubtrnka** [reply]
I'd suggest that we remove everything related to AES-GCM from the code. I think there is no mention of it in the spec anymore.
that multi-cipher feature turned out to be unnecessary.

**Reply by Fi3** [reply]
we still have it in the spec so we should also remove it also from there

**Reply by GitGab19** [reply]
> we still have it in the spec so we should also remove it also from there

Yeah AES-GCM is still there: https://github.com/stratum-mining/sv2-spec/blob/main/04-Protocol-Security.md#441-cipherstate-object
Opened an issue to remove it from specs: https://github.com/stratum-mining/sv2-spec/issues/93

**Reply by Sjors** [reply]
Where in the handshake could we even negotiate the choice between ChaCha20Poly1305 and AES-GCM? I think that was removed in #61.

On the Bitcoin Core side I haven't implemented the latter at all.

**Reply by Fi3** [reply]
yep we need to remove every reference to AES in the spec and also remove it from the impl

**Comment by bit-aloo** [comment]
@Fi3, Can you provide more insights on this code snippet? It makes sense for `Cipher` to be non-`Sync` and non-copyable so that we don't have multiple copies of it and to prevent concurrent access by multiple threads. However, I see that the implementation to enforce this is commented out. Why wasn't it implemented?
```
// Make sure that Cipher is not sync so we do not need to worry about what other memory accessor see
// after that we zeroize k is send cause if we send it the original thread can not access
// anymore it
//impl<C: AeadCipher> !Sync for Cipher<C> {}
//impl<C: AeadCipher> !Copy for Cipher<C> {}
impl<C: AeadCipher> Cipher<C> {
    /// Internal use only, we need k for handshake
    pub fn from_key_and_cipher(k: [u8; 32], c: C) -> Self {
        Self {
            k: Some(k),
            n: 0,
            cipher: Some(c),
        }
    }

    /// At the end of the handshake we return a cipher with hidden key
    #[allow(dead_code)]
    pub fn from_cipher(c: C) -> Self {
        Self {
            k: None,
            n: 0,
            cipher: Some(c),
        }
    }
}
``` 
cc: @rrybarczyk @plebhash

**Reply by Fi3** [reply]
yep it make sense at first look. But I really don't remember why I commented it out

**Reply by bit-aloo** [reply]
I tried uncommenting and building the code but encountered some errors. These errors suggest that 'negative impl' in Rust is a feature that is not yet fully stable or implemented.

**Reply by Fi3** [reply]
ok maybe it work only on nightly and I just commented it so when will be stabilized we can use it

**Reply by rrybarczyk** [reply]
Are there many features that are only supported on nightly? How can we tell which features are nightly only?

**Reply by bit-aloo** [reply]
Given the current structure of the `Cipher` struct, it is highly unlikely that it is `Sync` or `Copy`. Both the encryption key `k` and nonce `n` are primitives, making them `Sync` and `Copy`. However, the traits of the entire struct depend on all of its fields. The primary concern is the `cipher` field, which implements the `AeadCipher` trait. Since this trait includes methods that require a mutable reference to `self`, the `cipher` field is neither `Copy` nor `Sync`, which means the `Cipher` struct as a whole is also neither `Sync` nor `Copy`.

**Reply by Fi3** [reply]
@Shourya742 yep is correct, having it enforced by typesystem can useful if we modify the code. @rrybarczyk if it compile on stable we are not using features available only on nighlty

**Reply by rrybarczyk** [reply]
Sounds good, thank you.

**Comment by bit-aloo** [comment]
@Fi3, the generatekey method in handshake.rs, in this method we are generating key pairs and checking if the parity is odd or not, and if it is then we are generating a new pair. It is a recursive call. Why are we not just negating the key, in case when the parity is not what we expect? 

```rust
fn generate_key() -> Keypair {
        let secp = Secp256k1::new();
        let (secret_key, ) = secp.generate_keypair(&mut rand::thread_rng());
        let kp = Keypair::from_secret_key(&secp, &secret_key);
        if kp.x_only_public_key().1 == crate::PARITY {
            kp
        } else {
            Self::generate_key()
        }
    }
``` 
cc: @rrybarczyk

**Reply by Fi3** [reply]
negating the key is a far better solution

**Comment by plebhash** [comment]
what happens if we instantiate a `noise_sv2::Initiator` without a pk?

@Fi3

**Reply by Fi3** [reply]
double check it, but I think that in this case we do not check the certificate. So you still have an encrypted connection but you don't authenticate the upstream

**Reply by Fi3** [reply]
For example when you connect to a TP that you run locally

**Reply by bit-aloo** [reply]
For this case we dont check the static key of the responder in the certificate, and assume it to be correct.

**Reply by bit-aloo** [reply]
Why we even need this? What are the use cases when auth is not need and encryption does?

**Comment by plebhash** [comment]
this comment is confusing:

https://github.com/stratum-mining/stratum/blob/060131edeae2400e77476763d06809c73e86b2d1/protocols/v2/noise-sv2/src/cipher_state.rs#L196-L197

if this is for internal use only, why is this function public?

@Fi3

**Reply by Fi3** [reply]
don't know if you can't make it private without breaking other crates do it, remember to bump majior version if you do.

**Comment by rrybarczyk** [comment]
* Why do we publicly export this type from another crate:
https://github.com/stratum-mining/stratum/blob/e65cb320173b0048ec65dc4e042d8013a7fa962d/protocols/v2/noise-sv2/src/lib.rs#L6

* Instead of publicly exporting from another crate, these constants should be included in the `noise_sv2` crate instead of the `const_sv2` crate:
https://github.com/stratum-mining/stratum/blob/e65cb320173b0048ec65dc4e042d8013a7fa962d/protocols/v2/noise-sv2/src/lib.rs#L18

* Why do we publicly export the external `aes_gcm::aead::Error` type:
https://github.com/stratum-mining/stratum/blob/e65cb320173b0048ec65dc4e042d8013a7fa962d/protocols/v2/noise-sv2/src/lib.rs#L6
